### ISSUE TITLE
Simple Economy variation reveals mystery numerical errors

### ISSUE THREAD

**Issue by kurt-klingbeil** [NetLogo/NetLogo#1356]
[SimpleEcon_2_3.PDF](https://github.com/NetLogo/NetLogo/files/822075/SimpleEcon_2_3.PDF)

When the SimpleEconomy model was extended to include progressive taxation in addition to
a fixed flat tax, an interesting anomaly was encountered.
While this Economy scenario must be a fixed-sum-game since wealth is deducted from one turtle and passed to another turtle, this behaviour is only observed in the $5 flat tax scenario.
In the 5% scenario, the total wealth fluctuates and eventually diverges.

It is not obvious that any netlogo rules are being broken.

Curiously, the introduction of a temporary variable will eliminate this fluctuation/divergence.
Can someone identify the cause and whether this is a bug and how to avoid/eliminate this in general?
In this case, since it was known to be a fixed-sum-game, the error was obvious.
Other simulations might introduce the same kind of error which might go undetected.

Zeroing in to the essential trial-and-error-kids-in-the-hall-strategy code:

Broken code: March02 version
;; procedure to subtract "tax" and re-distribution
to give-away
  ask turtles  with [wealth > tax] [
    set wealth wealth - tax
    ask one-of other turtles [set wealth wealth + tax]
  ]
 end

Fixed code: March03 version
;; procedure to subtract "tax" and re-distribution
to give-away
  ask turtles  with [wealth > tax] [
    set wealth wealth - tax
    let xtemp tax
    ask one-of other turtles [set wealth wealth + xtemp]
  ]
 end

BTW, the attached PDF file is actually a ZIP file.
Attempts to drag&drop or select a SimpleEcon_2_3.zip (and SimpleEcon_2_3.ZIP) were rebuffed
so expediency demanded a PDF-impersonation which was accepted
So, additional questions are:
- whether zip attachments are indeed supported (or should be removed from the list), and 
- why .nlogo is NOT supported given that is the "native" extension/format of this code

**Comment by daimpi** [comment]
Are you aware of the fact that `xtemp` is not necessarily = `tax` ? Just from a quick look at your model, `tax` is a turtle-owned variable and therefore the `tax` you use might be different from `xtemp`, unless `tax` is the same for every turtle: in the broken code you first use the `tax` value of one turtle (lets say turtle 1) and then you ask another turtle (lets say turtle 2) and therefore use the turtles 2's `tax` variable which might might be different from the `tax` variable of turtle 1 which you seemingly intend to use.
In your "fixed code" you save the tax value of turtle 1 in a temporary variable `xtemp`, so if [tax] of (turtle 1) != [tax] of (turtle 2) the "broken code" is functional different from your "fixed code". 

Btw: stackoverflow might be a good place for such questions: http://stackoverflow.com/questions/tagged/netlogo

**Comment by mrerrormessage** [comment]
@kurt-klingbeil , sorry for the inability to post ".nlogo" files. GitHub controls file uploading and we don't have control over what file types or allowed. We *do* have users who post zip files as part of issues, so I'm not certain what went wrong in your case. It might be worth [contacting GitHub support](https://github.com/contact) to let them know and see if they have solutions.

My analysis of your model agrees with @daimpi . The two procedures you list above are not functionally equivalent. At the risk of being pedantic, I want to break down a bit more what is happening because the agent interaction in this case is actually quite subtle:

```NetLogo
to give-away-a
  ask turtles with [wealth > tax] [
    set wealth wealth - tax ; this code is run by some turtle (turtle 1), it subtract's turtle 1's tax from turtle 1's wealth
    ask one-of other turtles [
      set wealth wealth + tax ; this code is run by a different turtle (turtle 2), it adds turtle 2's tax to turtle 2's wealth
    ]
  ]
end

to give-away-b
  ask turtles with [wealth > tax] [
    set wealth wealth - tax ; this code is run by turtle 1, it subtract's turtle 1's tax from turtle 1's wealth
    let xtemp tax ; this code is run by turtle 1, it saves the value of turtle 1's tax to a local variable
    ask one-of other turtles [
      set wealth wealth + xtemp ; this code is run by turtle 2, it add's turtle 1's tax to turtle 2's wealth
    ]
  ]
end
```

One thing that you will notice is that in the first version of this code, the total wealth of all turtles (`sum [ wealth ] of turtles`) is only accidentally constant.

I think this would be a good place to address the concerns raised in #1357 as well. Again, at the risk of being pedantic, I want to get very precise about what NetLogo's intended behavior is. NetLogo's `ask` and `one-of` primitives appear to be similar to `while`-loops and `first` in other languages, but they're actually quite different. `ask` (and other agentset primitives) evaluate the block they are passed for each agent in a random (but reproducible) order. Similarly `one-of` returns a random (but reproducible) agent from the agentset it takes as an argument. This is actually a major component of NetLogo's design. From the [programming guide](https://ccl.northwestern.edu/netlogo/6.0/docs/programming.html#agentsets):

>  An agentset is not in any particular order. In fact, it's always in a random order. And every time you use it, the agentset is in a different random order. This helps you keep your model from treating any particular turtles, patches or links differently from any others (unless you want them to be). Since the order is random every time, no one agent always gets to go first. 

Because the agentset primitives are random, there are a combinatorial number of paths that this code might take (as there often are when working with agentsets). I'll enumerate these inductively starting with zero turtles and moving to two

1. There are no turtles with `wealth > tax`: nothing happens.
2. There is exactly one turtle with `wealth > tax`: the block is run for that turtle, it decreases its own wealth and picks exactly one other turtle at random and increases that turtle's wealth.
3. There are two turtles with `wealth > tax`: the block is run for each turtle, the following possible outcomes can occur based on the selection of `one-of`.
    1. Both turtles select a third and fourth turtle. Both of the turtles have their wealth decreased by their tax and the third and fourth turtles have their wealth increased: four turtles change their wealth.
    2. The first and second turtle both select the same third turtle. The first and second turtles have their wealth decreased by their tax and the third turtle has its wealth increased: three turtles change their wealth.
    3. One turtle selects the other turtle while the second selects some third turtle. Each of these turtles has its wealth increased. In the first version, the first turtle will have its wealth decreased by its tax, the third turtle will have its wealth increased by its own tax, and the second turtle will have its wealth unchanged: only two turtles change their wealth. In the second version, the first turtle will have its wealth decreased by its own tax, the second turtle will have its wealth changed by the difference between the first turtle's tax and its own tax, and the third turtle will have its wealth increased by the second turtles tax: only three turtles change their wealth (and only two if the two turtles start with the same wealth).
    4. The turtles select each other and increase the other's wealth. In the first version there are zero turtles which change their wealth. In second their are either two or zero turtles that change their wealth depending on whether or not the two turtles have the same wealth.

Because the possibilities are combinatoric, you will get *even more* possible outcomes when moving beyond two turtles with `wealth > tax` (and even wider swings in the number of turtles which appear to have changed their wealth, especially under the first procedure).

In summary, it seems that NetLogo is behaving as intended here and I will be closing this issue. If you believe that we've misunderstood something in your question, please let us know!

**Comment by kurt-klingbeil** [comment]
pedantry is great
thank you for your comprehensive explanation

I comprehend the variable scoping issue...
For the payor, their wealth is decreased by their tax rate while for the
receiver, their wealth is increased by their own tax rate (kinda like crony-corporatism)
Interposing the independent variable ensures the transfer occurs

The part about multiple interactions per step seems bizarre...
It makes sense for one transaction per step
I just can't grok Allowing all possible interactions per each step, 
or a random number of interactions per step.
That would be both an N-body effect and an acceleration of the simulation.
I notice that removing the "with [wealth > tax]" condition changes nothing.
